Use quarkus-operator-sdk-bundle-generator for OLM YAMLs

ASzc opened the pull request:
A work in progress. The plugin seems to be able to create the same files, except the CSV is missing some stuff. I will need to determine if there is some way to fix this, because the golang operator was able to generate the whole CSV from the code just fine.

Also it seems that I may need to create a new plugin, like I did a few years ago with jsonpath-maven-plugin, for YAML, so that some tweaks can be applied (like `replaces`) inside the maven build before the zip is assembled.

Closes #10911

ASzc committed:
Use quarkus-operator-sdk-bundle-generator for OLM YAMLs

Closes #10911

mabartos commented on operator/olm-base/manifests/clusterserviceversion.yaml:
```diff
@@ -1,143 +0,0 @@
-                      - name: QUARKUS_OPERATOR_SDK_NAMESPACES
-                        valueFrom:
-                          fieldRef:
-                            fieldPath: metadata.annotations['olm.targetNamespaces']
```
<img width="400" src="https://github.com/keycloak/keycloak/assets/38039883/1e8de0f1-5748-40bc-ba67-8283076720b3"/>

I've noticed the `QUARKUS_OPERATOR_SDK_NAMESPACES` env var was renamed to `KUBERNETES_NAMESPACE`. Can be like that? 

Also the value for the namespace is getting directly from the `metadata.namespace`, which I assume is directly specified in the `template.metadata.namespace` a few lines above and specified as 'placeholder', right? 

Does this [line](https://github.com/keycloak/keycloak/pull/20672/files#diff-f53c77df31edb884b75c19bd62de359748d617a94909dbafece4bbc435c0ee23R18) have any effect on this?

Pepo48 replied:
@mabartos I read about that on Monday too. Afaik you're correct - we can do it thanks to the values in application.properties file (see [this doc](https://quarkus.io/guides/deploying-to-kubernetes#namespace)).

But let's wait for Alex's confirmation.

ASzc replied:
Latest commit has manual yq commands to fix this

mabartos commented on operator/src/main/java/org/keycloak/operator/controllers/KeycloakSharedCsvMetadata.java:
```diff
@@ -0,0 +1,139 @@
+    },
+    maintainers = {
+        @CSVMetadata.Maintainer(
+            email = "keycloak-dev@lists.jboss.org",
```
```suggestion
            email = "keycloak-dev@googlegroups.com",
```
Regarding the Keycloak website, the official way is via Google groups.

ASzc replied:
I can update that for sure

shawkins commented:
After walking through the diff changes, they all seem correct / intentional. Just a couple of questions / thoughts:
- the inclusion of the app.kubernetes.io/version in the deployment template / match labels makes sense, just wanted to check if the intention is to eventually support multiple operator versions running concurrently?
- is adding resource requirements something that is planned?

ASzc commented:
@shawkins The `app.kubernetes.io/version` stuff is another hard coded behaviour of the bundle extension. There's no intention there from me specifically. If you think it will work, then ok. If not, I can ask the extension developer to add a feature to fix this.

For the resource requirements, I'm not 100% sure what you mean. If you're referring to the missing metadata under the `customresourcedefinitions:` part of the CSV, then the annotations I have added to the CRD classes will get used once bundle extension 6.1.2 is published. That's something that was missing from the extension, I had added.

ASzc commented:
I've rebased the branch, and fixed the mailing list email. My above comments are still open

ASzc commented:
@shawkins That sounds like a legitimate concern. I have no way to test the upgrade myself.

At this point it looks like I'm going to need to insert a lot of yq commands to make the CSV look like it did before. I really didn't want to do that, but I also can't let this PR sit any longer, it's very overdue.

ASzc commented:
I've pushed a commit with enough yq commands so that I'm confident that the CSVs are the same. New diff:
[csv-yaml-diff.txt](https://github.com/keycloak/keycloak/files/11791156/csv-yaml-diff.txt)

mabartos commented:
@ASzc Thanks for the update! Could you please review the test failures for Keycloak Operator CI / Test remote? Thanks!

shawkins commented:
LGTM I have no further concerns.

mabartos commented:
Same as @shawkins and @Pepo48. Once the CI pass, I'm ok with this PR.

@ASzc Thanks!

ASzc commented:
It appears the operator remote test is just timing out again, and will eventually fail. I have no idea why is this. The CSV is almost identical to before now. I will look into it

abstractj commented:
> It appears the operator remote test is just timing out again, and will eventually fail. I have no idea why is this. The CSV is almost identical to before now. I will look into it

@ASzc please ping me when ready, so we can merge it.

ASzc commented:
Tests are passing now! I think the main problem was one of the lines I put in the application.properties file. Since it's both build-time and run-time it was doing weird stuff with the basic CSV that the remote tests use (it doesn't use the normal one).
